lib/semian/configuration_validator.rb

```diff
@@ -0,0 +1,133 @@
+        raise ArgumentError, "Both bulkhead and circuitbreaker cannot be disabled."
+      end
+    end
+
+    def validate_bulkhead_configuration!
+      return if ENV.key?("SEMIAN_BULKHEAD_DISABLED")
+      return unless @configuration.fetch(:bulkhead, true)
+
+      tickets = @configuration[:tickets]
+      quota = @configuration[:quota]
+
+      if tickets.nil? && quota.nil?
+        raise ArgumentError, "Semian configuration require either the :tickets or :quota parameter, you provided neither"
+      end
+
+      if tickets && quota
+        raise ArgumentError, "Semian configuration require either the :tickets or :quota parameter, you provided both"
+      end
+
+      validate_quota!(quota) if quota
+      validate_tickets!(tickets) if tickets
+    end
+
+    def validate_circuit_breaker_configuration!
+      return if ENV.key?("SEMIAN_CIRCUIT_BREAKER_DISABLED")
+      return unless @configuration.fetch(:circuit_breaker, true)
+
+      require_keys!([:success_threshold, :error_threshold, :error_timeout], @configuration)
+      validate_thresholds!
+      validate_timeouts!
+    end
+
+    def validate_thresholds!
+      success_threshold = @configuration[:success_threshold]
+      error_threshold = @configuration[:error_threshold]
+
+      if success_threshold && success_threshold <= 0
+        raise ArgumentError, "success_threshold must be positive"
+      end
+
+      if error_threshold && error_threshold <= 0
+        raise ArgumentError, "error_threshold must be positive"
+      end
+    end
+
+    def validate_timeouts!
+      error_timeout = @configuration[:error_timeout]
+      error_threshold_timeout = @configuration[:error_threshold_timeout]
+      error_threshold = @configuration[:error_threshold]
+      lumping_interval = @configuration[:lumping_interval]
+      half_open_resource_timeout = @configuration[:half_open_resource_timeout]
+
+      unless error_timeout && error_timeout >= 0
+        raise ArgumentError, "error_timeout must be non-negative"
+      end
+
+      unless error_threshold_timeout.nil? || error_threshold_timeout >= 0
+        raise ArgumentError, "error_threshold_timeout must be non-negative"
+      end
+
+      unless half_open_resource_timeout.nil? || half_open_resource_timeout >= 0
+        raise ArgumentError, "half_open_resource_timeout must be non-negative"
+      end
+
+      unless lumping_interval.nil? || lumping_interval >= 0
+        raise ArgumentError, "lumping_interval must be non-negative"
+      end
+
+      unless lumping_interval.nil? || error_threshold_timeout.nil? || lumping_interval * (error_threshold - 1) <= error_threshold_timeout
+        raise ArgumentError, "constraint violated: lumping_interval * (error_threshold - 1) <= error_threshold_timeout, got lumping_interval: #{lumping_interval}, error_threshold: #{error_threshold}, error_threshold_timeout: #{error_threshold_timeout}"
+      end
+
+      unless half_open_resource_timeout.nil? || half_open_resource_timeout <= error_timeout
```

Comment:
Discussed in person: half_open does not care about thresholds. The half_open_resource_timeout should be less than the original resource timeout, but that's the only restriction